### ISSUE TITLE
Promise resolve/reject failure when invoked from a thread event handler

### ISSUE THREAD

**Issue by bobebert** [audreyt/node-webworker-threads#64]
With node stable (v0.12.7) and webworker-threads 0.5.6, calling a promise resolve or reject from a thread event handler fails silently: the promise is never completed.  One workaround is to resolve or reject the promise from a setTimeout callback.  Curiously, process.nextTick does not workaround the issue.

Example:

    "use strict";
    var Threads = require('webworker-threads');
    var thread = Threads.create();
    
    thread.eval(
        "thread.on('ping', function() {" +
            "print('got ping');" +
            "thread.emit('pong')" +
        "});");
    
    var p = new Promise(function(resolve, reject) {
        thread.on('pong', function() {
            console.log('got pong');
            resolve();                      // FAILS: promise is never resolved
            //process.nextTick(resolve);    // FAILS: promise is never resolved
            //setTimeout(resolve, 0);       // WORKS
    
            //reject();                     // FAILS: promise is never rejected
            //process.nextTick(reject);     // FAILS: promise is never rejected
            //setTimeout(reject, 0);        // WORKS
        });
        thread.emit('ping');
    });
    
    p.then(function() {
        console.log('resolved');
        process.exit(0);
    }).catch(function() {
        console.log('rejected');
        process.exit(1);
    });

**Comment by dmitrif** [comment]
+1

**Comment by simplyianm** [comment]
+1

**Comment by LeonineKing1199** [comment]
Okay, I went to le Reddit and I found another way of accomplishing this. Close the worker once you're done with it. Something like this :

```
'use strict';

const Threads = require('webworker-threads');
const Worker = Threads.Worker;

var p = new Promise((resolve, reject) => {
    var worker = new Worker(function(){
        function fibo (n) {
            return n > 1 ? fibo(n - 1) + fibo(n - 2) : 1;
        }

        this.onmessage = function (event) {
            console.log(`calculating fibonacci sequence of ${event.data}`)
            postMessage(fibo(event.data));
            self.close(); // this is important
        }
    });
  
    worker.onmessage = (event) => {
        resolve(event.data);
        worker.onmessage = null;
    };

    worker.postMessage(42);
});

p
    .then(console.log)
    .catch(err => console.log('err', err));
```

This is neat. You can't re-use the same Worker because once it's ```close()```d, it seems closed for good.

**Comment by rainder** [comment]
+1

**Comment by derMart** [comment]
yep, same here in 2020. Is there anyone maintaining this package anymore?
If not, anybody knows an alternative which gives the webworker api to node.js?

**Comment by addaleax** [comment]
@derMart In 2020, Node.js has [built-in Workers](https://nodejs.org/api/worker_threads.html). If you need something that matches the Web’s Worker API more closely, you can use e.g. https://github.com/developit/web-worker.

**Comment by derMart** [comment]
@addaleax thanks, yes that seems to be what I am looking for, just need the webworker API exactly as is in node.js

**Comment by derMart** [comment]
well anyway, if in 2020 this repo is deprecated, would be really nice to add that information to the README ;-)

**Comment by audreyt** [comment]
A pull request to deprecate this module and recommend https://github.com/developit/web-worker ) would be very much welcome!

**Comment by derMart** [comment]
sorry no time for that!
